packages/cli/src/util/dev/server.ts

```diff
@@ -1566,9 +1567,31 @@ export default class DevServer {
 
             // Retain orginal pathname, but override query parameters from the rewrite
             const beforeRewriteUrl = req.url || '/';
-            const rewriteUrlParsed = url.parse(beforeRewriteUrl);
-            rewriteUrlParsed.search = url.parse(rewritePath).search;
-            req.url = url.format(rewriteUrlParsed);
+
+            if (isURL(rewritePath)) {
+              const rewriteUrlParsed = new URL(rewritePath);
+
+              // `this.address` already has localhost normalized from ip4 and ip6 values
+              const devServerOrigin = new URL(this.address).origin;
+              const isSameOrigin = devServerOrigin === rewriteUrlParsed.origin;
+
+              if (isSameOrigin) {
```

Comment:
```suggestion
              const devServerParsed = new URL(this.address);
              if (rewriteUrlParsed.origin === devServerParsed.origin) {
```

Nit pick: remove a variable so its clear what is being compared and they aren't used later

Reply:
I find both roughly equal in clarity. I'd rather come through and update these references to `this.address` to be a new property that's already parsed or update `this.address` itself to already be parsed, as you mentioned below.

Reply:
I do like @styfle's recommendation. It's a bit easier to read as it has less "origins" that my mind has to process.

Reply:
Updated [`4bd0b08` (#8457)](https://github.com/vercel/vercel/pull/8457/commits/4bd0b08ace0308e37aa20b5573f0367a7cd6b9a2)